From 5edd79ae2c721a037d7c7b4c1de48a34b97fda9a Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 4 Jan 2017 15:34:43 +0800
Subject: [PATCH 1456/5242] MLK-13722-2 video: mxsfb: change 'usage' to
 atomic_t type

commit  a33c2aff5d27594a5ae7819d4db58a3aad24f730 from
https://source.codeaurora.org/external/imx/linux-imx.git

The 'usage' field of mxsfb_layer is used to record the
overlay fb user counts. So change its type to atomic_t
to avoid race problem.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index 3ca0737..6a6134a 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -56,6 +56,7 @@
 #include <linux/fb.h>
 #include <linux/mxcfb.h>
 #include <linux/regulator/consumer.h>
+#include <linux/types.h>
 #include <linux/videodev2.h>
 #include <video/of_display_timing.h>
 #include <video/videomode.h>
@@ -217,7 +218,7 @@ struct mxsfb_layer {
 	struct fb_info		*ol_fb;
 	int			id;
 	int			registered;
-	uint32_t		usage;
+	atomic_t		usage;
 	int			blank_state;
 	uint32_t 		global_alpha;
 
@@ -1662,7 +1663,7 @@ static int overlayfb_open(struct fb_info *info, int user)
 	struct mxsfb_layer *ofb = (struct mxsfb_layer*)info->par;
 	struct mxsfb_info  *fbi = ofb->fbi;
 
-	if (ofb->usage++ == 0) {
+	if (atomic_inc_return(&ofb->usage) == 1) {
 		memset((void*)&info->var, 0x0, sizeof(info->var));
 
 		ofb->ol_fb->var.xres		= fbi->fb_info->var.xres;
@@ -1681,9 +1682,9 @@ static int overlayfb_release(struct fb_info *info, int user)
 	struct mxsfb_layer *ofb = (struct mxsfb_layer*)info->par;
 	struct mxsfb_info *fbi = ofb->fbi;
 
-	BUG_ON(!ofb->usage);
+	BUG_ON(!atomic_read(&ofb->usage));
 
-	if (--ofb->usage == 0) {
+	if (atomic_dec_return(&ofb->usage) == 0) {
 		if (ofb->blank_state == FB_BLANK_UNBLANK)
 			ofb->ops->disable(ofb);
 
@@ -1906,7 +1907,7 @@ static void init_mxsfb_overlay(struct mxsfb_info *fbi,
 
 	ofb->id = 0;
 	ofb->ops = &ofb_ops;
-	ofb->usage = 0;
+	atomic_set(&ofb->usage, 0);
 	ofb->blank_state = -1;
 	ofb->global_alpha = 255;
 	ofb->fbi = fbi;
-- 
1.7.9.5

